### PR TITLE
ImageNet dataset

### PR DESCRIPTION
**Pull request by pmeier**
This is my first attempt to implement the `ImageNet` dataset as discussed in #713. I only used the official files, which can be downloaded [here](https://simon32.github.io/2018/01/09/image-net/).

Since the training and validation set as well as the meta information have to be downloaded separately I see no downside to structure the dataset properly. In anticipation of multiple years of the ILSVRC I've created the following dataset structure:
```
ILSVRC
└── 2012
    ├── meta.bin
    ├── train
        ├── n012345678
        ...
    └── val
        ├── n012345678
        ...
└── 2013
    ...
```
The sysnet identifier and the corresponding converter are accessible via the attributes `wnids` and `wnid_to_idx` while `classes` and `class_to_idx` now refer to the human-readable classes.

# Major Edits

- Within 616492e the attribute `year` was removed as requested. Thus, the tree now looks like this:
```
ILSVRC
├── meta.bin
├── train
    ├── n012345678
     ...
├── train.tar
├── val
    ├── n012345678
     ...
└── val.tar
```

# To Do

- ~~For now only the classification challenge is supported.~~
- ~~For now only the ILSVRC2012 is supported. The parsing of the development kit, which contains the meta information, is (probably) not yet applicable to other years.~~
- ~~I don't know, if the meta information is changing between the years. If not, I think it would be preferrable to have only one meta file in the `ILSVRC` folder.~~
- ~~The `class_to_idx` converter is not a good solution in its current state, since one requires the tuple of all correct class names to convert it to an index~~.
- In its current state, the meta file also contains the ground truth data of the validation set. I save this, since it is needed to prepare the validation folder. Afterwards this information is not needed anymore.
- ~~For now this is only tested on the validation set, since the download of the training set is still ongoing.~~

Let me know what you think.

### CODE REVIEW

**Comment by fmassa** [comment]
Hi,

One quick question: the `ImageNetDetection` that you mention is actually for classification, and not really for detection, right? I don't see mention of the bounding boxes for detection in your implementation

**Comment by pmeier** [comment]
@fmassa You are right. I misinterpreted the meaning of _detection_ and _segmentation_.

File: torchvision/datasets/imagenet.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+    def _load_meta(self):
+        # TODO: verify meta file
+        return torch.load(os.path.join(self.year_folder, META_FILE))[0]
+
+    def _prepare_tree(self):
+        try:
+            os.makedirs(self.split_folder)
+        except FileExistsError:
+            shutil.rmtree(self.split_folder)
```
Comment (fmassa) [inline torchvision/datasets/imagenet.py]: Are you sure we want to delete everything and re-download everything if we pass `download=True` again? For me we would just skip re-downloading it

File: torchvision/datasets/imagenet.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+META_FILE = 'meta.bin'
+
+
+class ImageNetClassification(ImageFolder):
+    def __init__(self, root, split='val', year='2012', download=False, **kwargs):
+
+        root = self.root = os.path.expanduser(root)
+        self.split = self._verify_split(split)
+        self.year = self._verify_year(year)
```
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/datasets/imagenet.py]: Can you clarify your intentions? Is this more "we have never thought about including the other years" or more "we will not use any ILSVRC other than 2012"? In the latter case I'm happy to throw it out.

File: torchvision/datasets/imagenet.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+ARCHIVE_DICT = {
+    ('2012', 'train'): {
+        'url': 'http://www.image-net.org/challenges/LSVRC/2012/nnoupb/ILSVRC2012_img_train.tar',
+        'md5': '1d675b47d978889d74fa0da5fadfb00e',
+    },
+    ('2012', 'val'): {
+        'url': 'http://www.image-net.org/challenges/LSVRC/2012/nnoupb/ILSVRC2012_img_val.tar',
+        'md5': '29b22e2961454d5413ddabcf34fc5622',
+    },
+    ('2012', 'devkit'): {
```
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/datasets/imagenet.py]: Correct.

File: torchvision/datasets/imagenet.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
+import os
+import sys
+if sys.version_info[0] == 2:
+    # FIXME: I don't know if this is good pratice / robust
+    FileExistsError = OSError
+
+import shutil
+import scipy.io as sio
```
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/datasets/imagenet.py]: Sure, we can do this. As stated above, I don't know if the meta information changes between the years. If we throw out the years completely, this shouldd be easy to handle.

File: torchvision/datasets/imagenet.py (hunk 1 above)
Comment (pmeier) [reply torchvision/datasets/imagenet.py]: I thought about that a bit. I intentionally defaulted the `download` flag to `False` so the user has to explicitly set it. If he does I don't see an option to check if the dataset is corrupted or not. In this case I check if the archive is available and download it if not. The code above only empties the `split_folder` which is for example `ILSVRC/2012/train`. This will be subsequently filled again by extracting the archives.

File: torchvision/datasets/imagenet.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    @property
+    def split_folder(self):
+        return os.path.join(self.year_folder, self.split)
+
+    @property
+    def archive_dict(self):
+        return ARCHIVE_DICT[(self.year, self.split)]
+
+
+class ImageNetDetection():
```
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/datasets/imagenet.py]: Sure, no problem.

File: torchvision/datasets/imagenet.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fmassa) [reply torchvision/datasets/imagenet.py]: I think I've never seen other years than 2012 being used, plus the classification challenge is (I think) not being updated anymore, so I think this is not necessary

File: torchvision/datasets/imagenet.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fmassa) [reply torchvision/datasets/imagenet.py]: moving it to inside the function is ok for now

File: torchvision/datasets/imagenet.py (hunk 1 above)
Comment (fmassa) [reply torchvision/datasets/imagenet.py]: The other datasets do not re-download the files if it's already available. I think this is particularly important for ImageNet, given it's size.

If the dataset is corrupted by some means, the user can himself delete the files and re-download maybe?

Comment (pmeier) [reply torchvision/datasets/imagenet.py]: My implementation also does not re-download the archives, if they are already present. Within the `download_and_extract_tar` the following lines achieve this:
```python
if not check_integrity(os.path.join(download_root, filename), md5):
    download_url(url, download_root, filename=filename, md5=md5)
```
Thus, it is only re-downloaded if
1. the archive is not present in the given root directory, or
2. its MD5 hash does not check out.

What I always do (if the user set `download=True`) is, deleting the contents of the `train` or `val` folder and re-extracting the archive. I do this, because without further information I'm not aware of a method to check the integrity of the extracted dataset.

With the development kit however (see discussion above) we could do some integrity checks:
1. Assert that `train` and `val` folder contain exactly `1000` subfolders with the correct wordnet IDs, and
2. assert that each subfolder contains the correct amount of images.

What do you think?

Comment (fmassa) [reply torchvision/datasets/imagenet.py]: Oh I see, thanks for the explanation!

This looks better to me (compared to re-downloading every time), but the side effect of removing the folders might be surprising.

Imagine for example that someone deleted the zips after the extraction was done, in order to save memory. Then, you would `rm` the whole `train` / `val` folders, requiring the user to re-download everything.

While I think that having more integrity checks is definitely a good thing, I'd rather do a single integrity check on the zips once its downloaded, than re-extract every time.

Thoughts?

File: torchvision/datasets/imagenet.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/datasets/imagenet.py]: Resolved in 616492e

File: torchvision/datasets/imagenet.py (hunk 1 above)
Comment (pmeier) [reply torchvision/datasets/imagenet.py]: I need some further clarification: How would your suggestion

> I'd rather do a single integrity check on the zips once its downloaded, than re-extract every time

solve the issue you described above? Without integrity checks of the extracted dataset, we have to re-download, if the user deletes the archives **and** sets `download=True`. Since I defaulted `download=False` this is a deliberate choice by the user and I see no way around  it.

Right now the code  does the following if the user sets `download=True`:
1. Empty the `split_folder`, i. e. delete all `train` or `val` images.
2. Check if the corresponding archive is available. If not, download it.
3. Extract the archive in the (now empty) `split_folder`.

The archive itself is **not** located inside the `split_folder` (take a look at the updated tree in the initial comment) and thus not deleted. IMO this is what you describe.

File: torchvision/datasets/imagenet.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/datasets/imagenet.py]: I've put some further thoughts into the parsing of the development kit. It mainly contains two things of importance to us:
1. The meta information, which allows us to convert between the WordNet IDs and a human readable description of each class, and
2. the ground truth for the classes of the validation images.

While the meta information is always needed, the ground truth is only needed once during the preparation of the `val` folder. I've created a [this gist](https://gist.github.com/pmeier/fd31c0bcb1769fb260ec34e84e00c1ed) as an example on how this could be stored in text files. If we do not include these files (or some that contain the same information) into the package, I don't see a way to _not_ parse the development kit.

File: torchvision/datasets/imagenet.py (hunk 1 above)
Comment (fmassa) [reply torchvision/datasets/imagenet.py]: If the user sets `download=True` and there is a folder existing with the expected name, then nothing would be downloaded. One other approach could be to store a dummy file, like `.torchvision_verified` in the root of the extracted dataset, so that it can be used by the verification logic if the zip file is not available.

Thoughts?

**Comment by fmassa** [comment]
I think this generally looks great, thanks!

I've made one more comment, and I'd have a question to you: what would be a good (or even the best) way to have a testing code to verify that the Dataset logic works fine?

Simply downloading the model file would be prohibitively expensive for large datasets (such as ImageNet). We could patch the download logic during testing, or have some small test files that would run during continuous integration.

I'd love to have your feedback here

File: torchvision/datasets/imagenet.py (hunk 1 above)
Comment (pmeier) [reply torchvision/datasets/imagenet.py]: Only addressing the first part of your comment here. For the second part see my answer below.

I think we are still talking past each other. At no point the code checks if a specific folder is present. I always _empty_ (recursively delete and create again) the `split_folder`. Thus, it doesn't matter whether the  `split_folder` exist or doesn't if the user sets `download=True`.

**Comment by pmeier** [comment]
I don't know if this is possible and feasible, but we could package our own _fake_ datasets. They should resemble the structure of the original dataset, but with drastically reduced number of instances (e. g. only one image per class for ImageNet). These could be downloaded and extracted quickly and thus a test could be run within continous integration.

But I don't think it is sufficient to check the download and extraction process this way even if we use the original dataset: no exception occuring during this process is IMO not a good metric to assert that the dataset is ready for usage afterwards. We should also check some statistics.

We could start off by creating dataset objects of all different combinations (mostly different `split`s but also `year`s for `VOC` etc.). For each of these objects we could check the following:

    - Is the number of instances correct?
    - Does the mean (or any other reasonable summary statistic) of all instances equal some pre-computed value?

If these checks pass, we can be sure that the download and extraction works correctly. However, we need to calculate the stored summary statistics without using the implemented procedure to avoid circular reasoning.

---

P.S.

On a second thought some variant of this could be also used verify the integrity of an `ImageFolder` dataset as `ImageNet`. While calculating the statistics for all instances is probably too time-consuming, applying this on a small subset (e. g. the first, the median, and the last) could already suffice.

**Comment by fmassa** [comment]
@pmeier I think this is in the right direction!

I think we don't need to have original dataset images to test the whole pipeline: only a set of randomly generated images of small size would be necessary.
For example, if we patch (using mock) `download_and_extract_tar` (or maybe even `tarfile`), we can test all the functionality of this dataset without having to download a single file.

While I agree that having integrity tests on the data downloaded would be nice to test, I think that it might actually make things more complicated: I'm not sure we can zip a few images from ImageNet due to licensing issues.

Here is an example of something that I think could be a nice inspiration: https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/image/imagenet_test.py and https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/testing/dataset_builder_testing.py
They have downloading logic for ImageNet, but they monkey-patch only a few functions so that everything can still be tested.

File: torchvision/datasets/imagenet.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+            prepare_val_folder(self.split_folder, val_wnids)
+
+    def _load_meta(self):
+        # TODO: verify meta file
+        return torch.load(os.path.join(self.root, META_DICT['filename']))[0]
+
+    def _empty_split_folder(self):
+        try:
+            shutil.rmtree(self.split_folder)
+        except FileNotFoundError:
```
Comment (fmassa) [inline torchvision/datasets/imagenet.py]: This is Python3 only unfortunately.

File: torchvision/datasets/imagenet.py (hunk 1 above)
Comment (fmassa) [reply torchvision/datasets/imagenet.py]: I still think that we should **not** delete the folder at first.

I think it might be more common than what you might imagine that people will go and delete the tar file containing the zipped images. They will be very surprised once they realize that they have to re-download everything because they deleted it.

Here is my point: downloading ImageNet is very time-consuming. If we want to perform some integrity check, we should do it just after downloading it. After we download it and the checks match, we assume that the zip extraction will work fine, and we don't need to worry about checking the hash all the time.

File: torchvision/datasets/imagenet.py (hunk 6 above)
Comment (pmeier) [reply torchvision/datasets/imagenet.py]: I've put in 
```python
if sys.version_info[0] == 2:
    # FIXME: I don't know if this is good pratice / robust
    FileNotFoundError = OSError
```
on top of the module. This worked for me in a quick test, but the `FIXME` note should be taken literally.

**Comment by pmeier** [comment]
Hm, you are right. I didn't think of the licensing issues. I've never worked with `mock` so it will take me some time to review the examples and come up with something. 

On a side note: should we move this discussion to another issue, since it applies to all datasets rather than only ImageNet?

File: torchvision/datasets/imagenet.py (hunk 1 above)
Comment (pmeier) [reply torchvision/datasets/imagenet.py]: > I still think that we should **not** delete the folder **at first**. (additional emphasis added)

Is it about the order of the operation, as in check if we have the archive before we empty the folder? If so, we could simply move the `_empty_split_folder()` call just before the extraction. This makes more sense now, since before the function created the whole tree for the `year`s etc.

> we don't need to worry about checking the hash all the time.

All this checking is only done if the user explicitly sets `download=True` which normally defaults to `False`. If I understand you correctly, all these issues can only occur if the dataset was downloaded before and the user creates the object with `download=True`. Do you want the default to be `download=True` and _protect_ the user in case he deleted the archives? Otherwise I still don't understand your point.

Comment (fmassa) [reply torchvision/datasets/imagenet.py]: My point is that the user has a single code, written `download=True`, and has run it for the first time and it worked.
Then, he opened the download folder and realized that the zip was still there. He deleted the zip because it take a bunch of space, and then run the same code again, using `download=True` because he hasn't changed it.

Then, he realized that the folder got erased, and it starts to download the tar again!

**Comment by fmassa** [comment]
@pmeier yes, the testing is something that should be done for all datasets.

I'll start working on it right now. I believe this is a pretty high-pri feature to have right now, as it is very difficult to review dataset code and ensure that it works as expected for both Py2 and Py3.

I'll try to push some POC implementation for the tests today and I'll tag you there

File: torchvision/datasets/imagenet.py (hunk 1 above)
Comment (pmeier) [reply torchvision/datasets/imagenet.py]: Finally got it. Do you want some logic with a verification file or is a check if the folder is present sufficient?

**Comment by pmeier** [comment]
@fmassa Any progress?

**Comment by fmassa** [comment]
@pmeier I didn't have the time yet to finish the tests for the datasets: this involved refactoring the downloading abstractions and I got stuck with other things. I'll add tests for this PR after merging it.
